src/machines/java/com/enderio/machines/common/blockentity/SoulBinderBlockEntity.java

```diff
@@ -0,0 +1,146 @@
+import com.enderio.machines.common.menu.SoulBinderMenu;
+import com.enderio.machines.common.recipe.SoulBindingRecipe;
+import net.minecraft.core.BlockPos;
+import net.minecraft.nbt.CompoundTag;
+import net.minecraft.world.entity.player.Inventory;
+import net.minecraft.world.entity.player.Player;
+import net.minecraft.world.inventory.AbstractContainerMenu;
+import net.minecraft.world.level.block.entity.BlockEntityType;
+import net.minecraft.world.level.block.state.BlockState;
+import net.minecraftforge.fluids.FluidStack;
+import net.minecraftforge.fluids.capability.IFluidHandler;
+import net.minecraftforge.fluids.capability.templates.FluidTank;
+import org.jetbrains.annotations.Nullable;
+
+import java.util.List;
+
+import static com.enderio.base.common.util.ExperienceUtil.EXPTOFLUID;
+
+public class SoulBinderBlockEntity extends PoweredCraftingMachine<SoulBindingRecipe, SoulBindingRecipe.Container> {
+
+    public static final QuadraticScalable CAPACITY = new QuadraticScalable(CapacitorModifier.ENERGY_CAPACITY, () -> 100000f);
+    public static final QuadraticScalable TRANSFER = new QuadraticScalable(CapacitorModifier.ENERGY_TRANSFER, () -> 120f);
+    public static final QuadraticScalable USAGE = new QuadraticScalable(CapacitorModifier.ENERGY_USE, () -> 30f);
+    private final SoulBindingRecipe.Container container;
+    public static final SingleSlotAccess INPUT_SOUL = new SingleSlotAccess();
+    public static final SingleSlotAccess INPUT_OTHER = new SingleSlotAccess();
+    public static final MultiSlotAccess OUTPUT = new MultiSlotAccess();
+    private final FluidTank fluidTank;
+    private final MachineFluidHandler fluidHandler;
+
+    public SoulBinderBlockEntity(BlockEntityType<?> type, BlockPos worldPosition, BlockState blockState) {
+        super(MachineRecipes.SOUL_BINDING.type().get(), CAPACITY, TRANSFER, USAGE, type, worldPosition, blockState);
+
+        fluidTank = createFluidTank(10000); //Todo EXP tag?
+        container = new SoulBindingRecipe.Container(getInventory(), fluidTank);
+
+        // Create fluid tank storage.
+        this.fluidHandler = new MachineFluidHandler(getIOConfig(), fluidTank);
+        addCapabilityProvider(fluidHandler);
+
+        addDataSlot(new IntegerDataSlot(() -> fluidTank.getFluidInTank(0).getAmount(), (i) -> fluidTank.setFluid(new FluidStack(EIOFluids.XP_JUICE.get(), i)),
+            SyncMode.WORLD));
+        addDataSlot(new IntegerDataSlot(() -> container.getNeededXP(), (i) -> container.setNeededXP(i), SyncMode.GUI));
+    }
+
+    @Override
+    public AbstractContainerMenu createMenu(int pContainerId, Inventory pPlayerInventory, Player pPlayer) {
+        return new SoulBinderMenu(this, pPlayerInventory, pContainerId);
+    }
+
+    @Override
+    public MachineInventoryLayout getInventoryLayout() {
+        return MachineInventoryLayout.builder()
+            .setStackLimit(1)
+            .inputSlot((slot, stack) -> stack.getCapability(EIOCapabilities.ENTITY_STORAGE).isPresent())
```

Comment:
This capability is present in both the soul vial and broken spawner. The recipe only allows for soul vials in the first position

Reply:
You can also add Powered Spawners in the first slot which shouldn't happen

Reply:
The recipe checks for soul vial and so it probably shouldn't accept other items (since we don't a recipe validity notifier inside the screen). I can see people getting confused when items get accepted in slots but nothing happens.

Reply:
I wanted to be future proof in case we did want a recipe that used something else. But now that I think about it, that's probably not gonna happen. I'll change it to only take soul vials, that's probably fine.